crates/client/mempool/src/lib.rs

```diff
@@ -0,0 +1,193 @@
+    #[error(transparent)]
+    InnerMempool(#[from] inner::TxInsersionError),
+    #[error(transparent)]
+    Exec(#[from] dc_exec::Error),
+}
+
+pub struct Mempool {
+    backend: Arc<DeoxysBackend>,
+    l1_data_provider: Arc<dyn L1DataProvider>,
+    inner: RwLock<MempoolInner>,
+}
+
+impl Mempool {
+    pub fn new(backend: Arc<DeoxysBackend>, l1_data_provider: Arc<dyn L1DataProvider>) -> Self {
+        Mempool { backend, l1_data_provider, inner: Default::default() }
+    }
+
+    /// This function creates the pending block if it is not found.
+    // TODO: move this somewhere else
+    pub(crate) fn get_or_create_pending_block(&self) -> Result<DeoxysMaybePendingBlock, Error> {
+        match self.backend.get_block(&DbBlockId::Pending)? {
+            Some(block) => Ok(block),
+            None => {
+                // No pending block: we create one :)
+
+                let block_info =
+                    self.backend.get_block_info(&BlockId::Tag(BlockTag::Latest))?.ok_or(Error::NoGenesis)?;
+                let block_info = block_info.as_nonpending().ok_or(Error::Internal("Latest block is pending".into()))?;
+
+                Ok(DeoxysMaybePendingBlock {
+                    info: DeoxysMaybePendingBlockInfo::Pending(DeoxysPendingBlockInfo {
+                        header: PendingHeader {
+                            parent_block_hash: block_info.block_hash,
+                            sequencer_address: **self.backend.chain_config().sequencer_address,
+                            block_timestamp: SystemTime::now()
+                                .duration_since(SystemTime::UNIX_EPOCH)
+                                .expect("Current time is before the unix timestamp")
+                                .as_secs(),
+                            protocol_version: self.backend.chain_config().latest_protocol_version,
+                            l1_gas_price: self.l1_data_provider.get_gas_prices(),
+                            l1_da_mode: self.l1_data_provider.get_da_mode(),
+                        },
+                        tx_hashes: vec![],
+                    }),
+                    inner: DeoxysBlockInner { transactions: vec![], receipts: vec![] },
+                })
+            }
+        }
+    }
+
+    pub fn accept_account_tx(&self, tx: AccountTransaction) -> Result<(), Error> {
+        // The timestamp *does not* take the transaction validation time into account.
+        let arrived_at = ArrivedAtTimestamp::now();
+
+        // Get pending block
+        let pending_block_info = self.get_or_create_pending_block()?;
+
+        // If the contract has been deployed for the same block is is invoked, we need to skip validations.
+        // NB: the lock is NOT taken the entire time the tx is being validated. As such, the deploy tx
+        //  may appear during that time - but it is not a problem.
+        let deploy_account_tx_hash = if let AccountTransaction::Invoke(tx) = &tx {
```

Comment:
I think a Declare is also possible here.

Reply:
> I think a Declare is also possible here.

I don't think the starkware sequencer supports using classes before they are declared?
It's just for deploy account and invoke I think

Reply:
So I was actually saying that it's possible that 

Nonce 0 - Deploy account
Nonce 1 - Declare some random class (we should ignore validation)

And the above check ignores validation only if Nonce 1 is an invoke if I understand it correctly, so the above case will be missed right?

Reply:
Yes, I believe this is what blockifier expects us to do
The predicate for validation skipping in is really hard coded to be an invoke after a deploy_account in blockifier